src/libmongoc/src/mongoc/mcd-rpc.h

```diff
@@ -0,0 +1,776 @@
+#define MONGOC_OP_CODE_QUERY INT32_C (2004)
+#define MONGOC_OP_CODE_GET_MORE INT32_C (2005)
+#define MONGOC_OP_CODE_DELETE INT32_C (2006)
+#define MONGOC_OP_CODE_KILL_CURSORS INT32_C (2007)
+
+#define MONGOC_OP_REPLY_RESPONSE_FLAG_NONE INT32_C (0)
+#define MONGOC_OP_REPLY_RESPONSE_FLAG_CURSOR_NOT_FOUND (INT32_C (0x01) << 0)
+#define MONGOC_OP_REPLY_RESPONSE_FLAG_QUERY_FAILURE (INT32_C (0x01) << 1)
+#define MONGOC_OP_REPLY_RESPONSE_FLAG_SHARD_CONFIG_STALE (INT32_C (0x01) << 2)
+#define MONGOC_OP_REPLY_RESPONSE_FLAG_AWAIT_CAPABLE (INT32_C (0x01) << 3)
+
+#define MONGOC_OP_UPDATE_FLAG_NONE INT32_C (0)
+#define MONGOC_OP_UPDATE_FLAG_UPSERT (INT32_C (0x01) << 0)
+#define MONGOC_OP_UPDATE_FLAG_MULTI_UPDATE (INT32_C (0x01) << 1)
+
+#define MONGOC_OP_INSERT_FLAG_NONE INT32_C (0)
+#define MONGOC_OP_INSERT_FLAG_CONTINUE_ON_ERROR (INT32_C (0x01) << 0)
+
+#define MONGOC_OP_QUERY_FLAG_NONE INT32_C (0)
+#define MONGOC_OP_QUERY_FLAG_TAILABLE_CURSOR (INT32_C (0x01) << 1)
+#define MONGOC_OP_QUERY_FLAG_SECONDARY_OK (INT32_C (0x01) << 2)
+#define MONGOC_OP_QUERY_FLAG_OPLOG_REPLAY (INT32_C (0x01) << 3)
+#define MONGOC_OP_QUERY_FLAG_NO_CURSOR_TIMEOUT (INT32_C (0x01) << 4)
+#define MONGOC_OP_QUERY_FLAG_AWAIT_DATA (INT32_C (0x01) << 5)
+#define MONGOC_OP_QUERY_FLAG_EXHAUST (INT32_C (0x01) << 6)
+#define MONGOC_OP_QUERY_FLAG_PARTIAL (INT32_C (0x01) << 7)
+
+#define MONGOC_OP_DELETE_FLAG_NONE INT32_C (0)
+#define MONGOC_OP_DELETE_FLAG_SINGLE_REMOVE (INT32_C (0x01) << 0)
+
+
+// Convert the given array of bytes into an RPC message object. The RPC message
+// object must be freed by `mcd_rpc_message_destroy`.
+//
+// data: an array of `length` bytes.
+// data_end: if not `NULL`, `*data_end` is set to one past the last byte of
+//           valid input data. Useful for diagnosing failures.
+//
+// Note: the fields of the RPC message object are automatically converted from
+// little endian to native endian.
+//
+// Returns the new RPC message object on success. Returns `NULL` on failure.
+mcd_rpc_message *
+mcd_rpc_message_from_data (const void *data,
+                           size_t length,
+                           const void **data_end);
+
+// The in-place version of `mcd_rpc_message_from_data`.
+//
+// rpc: an RPC message object in an initialized state.
+//
+// Returns `true` on success. Returns `false` on failure.
+bool
+mcd_rpc_message_from_data_in_place (mcd_rpc_message *rpc,
+                                    const void *data,
+                                    size_t length,
+                                    const void **data_end);
+
+// Convert the given RPC message object into an array of iovec structures. The
+// return value must be freed by `bson_free`.
+//
+// rpc: a valid RPC message object whose fields are in native endian.
+// length: if not `NULL`, `*length` is set to the number of iovec structures in
+//         the array.
+//
+// Note: this function converts the fields of the RPC message object to little
+// endian.
```

Comment:
```suggestion
// endian. Caller must not call more than once. Caller is expected not to modify or inspect `rpc` after calling.
```

If a caller calls `mcd_rpc_message_to_iovecs` twice on a big-endian machine, I expect the fields would swap back the bytes back to big-endian on the second call. A bug like that may not be realized until an application runs on a big-endian machine.

Once `mcd_rpc_message_to_iovecs` is called, calls to get fields (e.g. `mcd_rpc_op_get_more_get_number_to_return`) would return the little-endian representation.

Consider: add a bool to track whether the fields have been converted to little-endian:

```c
struct _mcd_rpc_message_header {
   int32_t message_length;
   int32_t request_id;
   int32_t response_to;
   int32_t op_code;
   bool converted_to_le; // Set to true when `mcd_rpc_message_to_iovecs` is called.
};
```

And assert `!converted_to_le` in `mcd_rpc_message_to_iovecs` and other methods that return or set fields.

Reply:
I wondered whether or not it would be worth asserting the "no-longer-native endian" state. It is effectively already being done by the `opCode` precondition checks for opcode functions, but as you say, that would not detect incorrect code until it is actually run on a big-endian target. Will add the converted state tracker as suggested.